modules/ttdBidAdapter.js

```diff
@@ -0,0 +1,602 @@
+  }
+  if (bidderRequest.uspConsent) {
+    utils.deepSetValue(regs, 'ext.us_privacy', bidderRequest.uspConsent);
+  }
+  if (config.getConfig('coppa') === true) {
+    regs.coppa = 1;
+  }
+  return regs;
+}
+
+function getBidFloor(bid) {
+  if (!utils.isFn(bid.getFloor)) {
+    return null;
+  }
+
+  let floor = bid.getFloor({
+    currency: 'USD',
+    mediaType: '*',
+    size: '*'
+  });
+  if (utils.isPlainObject(floor) && !isNaN(floor.floor) && floor.currency === 'USD') {
+    return floor.floor;
+  }
+  return null;
+}
+
+function getSource(validBidRequests) {
+  let source = {};
+  if (validBidRequests[0].schain) {
+    utils.deepSetValue(source, 'ext.schain', validBidRequests[0].schain);
+  }
+  return source;
+}
+
+function getDevice() {
+  const language = navigator.language || navigator.browserLanguage || navigator.userLanguage || navigator.systemLanguage;
+  let device = {
+    ua: navigator.userAgent,
+    dnt: utils.getDNT() ? 1 : 0,
+    language: language,
+    connectiontype: getConnectionType()
+  };
+
+  return device;
+};
+
+function getConnectionType() {
+  const connection = navigator.connection || navigator.webkitConnection;
+  if (!connection) {
+    return 0;
+  }
+  switch (connection.type) {
+    case 'ethernet':
+      return 1;
+    case 'wifi':
+      return 2;
+    case 'cellular':
+      switch (connection.effectiveType) {
+        case 'slow-2g':
+        case '2g':
+          return 4;
+        case '3g':
+          return 5;
+        case '4g':
+          return 6;
+        default:
+          return 3;
+      }
+    default:
+      return 0;
+  }
+}
+
+function getUser(bidderRequest) {
+  let user = {};
+  if (bidderRequest.gdprConsent) {
+    utils.deepSetValue(user, 'ext.consent', bidderRequest.gdprConsent.consentString);
+  }
+
+  if (utils.isStr(utils.deepAccess(bidderRequest, 'bids.0.userId.tdid'))) {
+    user.buyeruid = bidderRequest.bids[0].userId.tdid;
+  }
+
+  var eids = createEidsArray(utils.deepAccess(bidderRequest, 'bids.0.userId'))
+  if (eids.length) {
+    utils.deepSetValue(user, 'ext.eids', eids);
+  }
+
+  return user;
+}
+
+function getSite(bidderRequest) {
+  var site = {
+    id: utils.deepAccess(bidderRequest, 'bids.0.params.siteId'),
+    page: utils.deepAccess(bidderRequest, 'refererInfo.referer'),
+    publisher: {
+      id: utils.deepAccess(bidderRequest, 'bids.0.params.publisherId'),
+    },
+    pagecat: utils.deepAccess(bidderRequest, 'bids.0.params.categories')
+  };
+
+  var publisherDomain = config.getConfig('publisherDomain');
+  if (publisherDomain) {
+    utils.deepSetValue(site, 'publisher.domain', publisherDomain);
+  }
+  return site;
+}
+
+function getImpression(bid) {
+  let impression = {
+    id: bid.bidId,
+    tagid: bid.params.placementId
+  };
+
+  if (bid.params.secure) {
+    impression.secure = bid.params.secure;
+  }
+
+  let gpid = utils.deepAccess(bid, 'ortb2Imp.ext.data.pbadslot');
```

Comment:
this is no longer the correct place to find the gpid see #7794 and #7671

Reply:
Fixed - gpid is now read from `ortb2Imp.ext.gpid` instead